src/server/engine_factory.rs

```diff
@@ -129,172 +105,125 @@ impl KvEngineFactory {
+        db_opts.set_env(self.inner.env.clone());
+        db_opts.add_event_listener(RocksEventListener::new(
             "kv",
             self.inner.sst_recovery_sender.clone(),
         ));
         if let Some(filter) = self.create_raftstore_compaction_listener() {
-            kv_db_opts.add_event_listener(filter);
+            db_opts.add_event_listener(filter);
         }
-        if let Some(listener) = &self.inner.flow_listener {
-            kv_db_opts.add_event_listener(listener.clone_with(region_id, suffix));
-        }
-        let kv_cfs_opts = self.inner.rocksdb_config.build_cf_opts(
+        db_opts
+    }
+
+    fn cf_opts(&self) -> Vec<(&str, RocksCfOptions)> {
+        self.inner.rocksdb_config.build_cf_opts(
             &self.inner.block_cache,
             self.inner.region_info_accessor.as_ref(),
             self.inner.api_version,
-        );
-        let kv_engine = engine_rocks::util::new_engine_opt(
-            tablet_path.to_str().unwrap(),
-            kv_db_opts,
-            kv_cfs_opts,
-        );
+        )
+    }
+
+    /// Create a shared db.
+    ///
+    /// It will always create in path/DEFAULT_DB_SUB_DIR.
+    pub fn create_shared_db(&self, path: &Path) -> Result<RocksEngine> {
+        let mut db_opts = self.db_opts();
+        let cf_opts = self.cf_opts();
+        if let Some(listener) = &self.inner.flow_listener {
+            db_opts.add_event_listener(listener.clone());
+        }
+        let target_path = path.join(DEFAULT_ROCKSDB_SUB_DIR);
+        let kv_engine =
+            engine_rocks::util::new_engine_opt(target_path.to_str().unwrap(), db_opts, cf_opts);
         if let Err(e) = &kv_engine {
-            error!("failed to create kv engine"; "path" => %tablet_path.display(), "err" => ?e);
+            error!("failed to create kv engine"; "path" => %path.display(), "err" => ?e);
         }
         kv_engine
     }
+}
 
-    pub fn on_tablet_created(&self, region_id: u64, suffix: u64) {
-        if let Some(listener) = &self.inner.flow_listener {
-            let listener = listener.clone_with(region_id, suffix);
-            listener.on_created();
+impl TabletFactory<RocksEngine> for KvEngineFactory {
+    fn open_tablet(&self, id: u64, suffix: Option<u64>, path: &Path) -> Result<RocksEngine> {
+        let mut db_opts = self.db_opts();
+        let cf_opts = self.cf_opts();
+        if let Some(listener) = &self.inner.flow_listener && let Some(suffix) = suffix {
+            db_opts.add_event_listener(listener.clone_with(id, suffix));
         }
+        let kv_engine =
+            engine_rocks::util::new_engine_opt(path.to_str().unwrap(), db_opts, cf_opts);
+        if let Err(e) = &kv_engine {
+            error!("failed to create tablet"; "id" => id, "suffix" => ?suffix, "path" => %path.display(), "err" => ?e);
+        } else if let Some(listener) = &self.inner.flow_listener && let Some(suffix) = suffix {
+            listener.clone_with(id, suffix).on_created();
```

Comment:
on_created should be put in TabletRegistry. Otherwise the consumer (TabletFlowController) will try to get it from TabletRegistry, but failed because it may not be added to registry yet.